pkg/controller/networkpolicy/clustergroup.go

```diff
@@ -333,25 +365,48 @@ func (n *NetworkPolicyController) GetAssociatedGroups(name, namespace string) ([
 	if !exists {
 		return nil, nil
 	}
-	groupObjs := make([]antreatypes.Group, 0, len(clusterGroups))
+	var groupObjs []antreatypes.Group
 	for _, g := range clusterGroups {
-		group, found, _ := n.internalGroupStore.Get(g)
-		if found {
-			groupObjs = append(groupObjs, *group.(*antreatypes.Group))
+		groupObjs = append(groupObjs, n.getAssociatedGroupsByName(g)...)
+	}
+	// Remove duplicates in the groupObj slice.
+	groupKeys, j := make(map[string]bool), 0
+	for _, g := range groupObjs {
+		if _, exists := groupKeys[g.Name]; !exists {
+			groupKeys[g.Name] = true
+			groupObjs[j] = g
+			j++
 		}
 	}
-	return groupObjs, nil
+	return groupObjs[:j], nil
+}
+
+// getAssociatedGroupsByName retrieves the internal Group and all it's parent Group objects
+// (if any) by Group name. This function assumes that there's no loop in Group reference.
+func (n *NetworkPolicyController) getAssociatedGroupsByName(grpName string) []antreatypes.Group {
+	var groups []antreatypes.Group
+	groupObj, found, _ := n.internalGroupStore.Get(grpName)
+	if found {
+		grp := groupObj.(*antreatypes.Group)
```

Comment:
If we handle not found case first, there could be less indent overall. This applies to other functions in this file too.

Reply:
Done